src/main/java/org/openstreetmap/atlas/utilities/collections/ShardBucketCollection.java

```diff
@@ -0,0 +1,454 @@
+import java.lang.reflect.Array;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.openstreetmap.atlas.geography.Located;
+import org.openstreetmap.atlas.geography.Rectangle;
+import org.openstreetmap.atlas.geography.index.RTree;
+import org.openstreetmap.atlas.geography.sharding.Shard;
+import org.openstreetmap.atlas.geography.sharding.Sharding;
+import org.openstreetmap.atlas.geography.sharding.SlippyTile;
+import org.openstreetmap.atlas.utilities.scalars.Counter;
+
+/**
+ * A collection wrapper for a set of collections associated with shards containing located items,
+ * such as CheckFlags or AtlasEntities. Supports concurrent add, contains, remove. The term bucket
+ * is used because of the usability of the collections mostly as storage for various location based
+ * sorting tasks. Does not support safe iteration while modifying.
+ *
+ * @param <LocatedType>
+ *            Located item type
+ * @param <CollectionType>
+ *            Type of the Collection of LocatedType associated with shards
+ * @author jklamer
+ */
+public abstract class ShardBucketCollection<LocatedType extends Located & Serializable, CollectionType extends Collection<LocatedType> & Serializable>
+        implements Collection<LocatedType>, Serializable
+{
+
+    /**
+     * A helper class that associates a shard with the index that its collection is at in the
+     * collectionBuckets array.
+     */
+    private static class ShardToCollectionIndex implements Located
+    {
+        private static final long serialVersionUID = 4050100671815503794L;
+        private final int index;
+        private final Shard shard;
+
+        ShardToCollectionIndex(final int index, final Shard shard)
+        {
+            this.index = index;
+            this.shard = shard;
+        }
+
+        @Override
+        public Rectangle bounds()
+        {
+            return this.getShard().bounds();
+        }
+
+        public int getIndex()
+        {
+            return this.index;
+        }
+
+        public Shard getShard()
+        {
+            return this.shard;
+        }
+    }
+
+    private static final long serialVersionUID = -7892704554302160820L;
+
+    private final CollectionType[] collectionBuckets;
+    private final RTree<ShardToCollectionIndex> collectionIndex;
+    private final HashMap<Shard, ShardToCollectionIndex> initializedShards = new HashMap<>();
+    private Rectangle maximumBounds;
+
+    public ShardBucketCollection(final Rectangle maximumBounds, final Integer zoomLevel)
+    {
+        this(maximumBounds, SlippyTile.allTiles(zoomLevel, maximumBounds));
+    }
+
+    public ShardBucketCollection(final Rectangle maximumBounds, final Sharding sharding)
```

Comment:
Sharding implementations are already behaving like a spatial index. I believe here you could gain some performance and simplification by using the sharding itself to query the intersecting bounds instead of a RTree, as the scope here is much smaller that what RTrees can handle...

Reply:
Well, if I query the sharing directly I would also need a map of of shard -> index, so that would require two look-ups instead of just the one now, even if the second look up is constant time it is more steps. I would also need to store both sharding and the map of shard ->index, instead of just the RTree.

 In terms of what RTree's can handle, this allows for initializations for slippy tiles up to the SlippyTile max zoom, which is 30. At zoom 19 there are [274.9 billion tiles](https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Zoom_levels). So even if the max bounds isn't the whole world the Rtree needs the ability to hold a lot even if my tests didn't work it that hard.

Reply:
I see, if it works for you now, we can merge it!

Reply:
😄